File: app/src/main/java/br/edu/ifspsaocarlos/agenda/activity/BaseActivity.java
package br.edu.ifspsaocarlos.agenda.activity;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toast;

import java.util.List;

import br.edu.ifspsaocarlos.agenda.R;
import br.edu.ifspsaocarlos.agenda.adapter.ContatoArrayAdapter;
import br.edu.ifspsaocarlos.agenda.data.ContatoDAO;
import br.edu.ifspsaocarlos.agenda.model.Contato;

public class BaseActivity extends AppCompatActivity {

    protected ContatoDAO cDAO=new ContatoDAO(this);
    public ListView list;
    public ContatoArrayAdapter adapter;
    protected SearchView searchView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        list = (ListView) findViewById(R.id.listView);
        list.setOnItemClickListener(new AdapterView.OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> adapterView, View arg1, int arg2,
                                    long arg3) {
                Contato contact = (Contato) adapterView.getAdapter().getItem(arg2);
                Intent inte = new Intent(getApplicationContext(), DetalheActivity.class);
                inte.putExtra("contact", contact);
                startActivityForResult(inte, 0);

            }

        });


        registerForContextMenu(list);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        getMenuInflater().inflate(R.menu.menu_main, menu);

        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchView = (SearchView) menu.findItem(R.id.pesqContato).getActionView();

        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));

        searchView.setIconifiedByDefault(true);
        return true;
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v,
                                    ContextMenu.ContextMenuInfo menuInfo) {
        // TODO Auto-generated method stub
        super.onCreateContextMenu(menu, v, menuInfo);
        MenuInflater m = getMenuInflater();
        m.inflate(R.menu.menu_context, menu);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {

        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();
        ContatoArrayAdapter adapter = (ContatoArrayAdapter)list.getAdapter();
        Contato contact = adapter.getItem(info.position);


        switch(item.getItemId()){
            case R.id.delete_item:
                cDAO.deleteContact(contact);
                Toast.makeText(getApplicationContext(), "Removido com sucesso", Toast.LENGTH_SHORT).show();
                buildListView();
                return true;
        }
        return super.onContextItemSelected(item);
    }


    protected void buildListView() {
        List<Contato> values = cDAO.buscaTodosContatos();
        adapter = new ContatoArrayAdapter(this, values);
        list.setAdapter(adapter);

    }

    protected void buildSearchListView(String query) {
        List<Contato> values = cDAO.buscaContato(query);
        adapter= new ContatoArrayAdapter(this, values);
        list.setAdapter(adapter);
    }

}

File: app/src/main/java/br/edu/ifspsaocarlos/agenda/activity/BaseActivity.kt
package br.edu.ifspsaocarlos.agenda.activity

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.SearchView
import android.support.v7.widget.Toolbar
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView.AdapterContextMenuInfo
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import android.widget.Toast
import br.edu.ifspsaocarlos.agenda.R
import br.edu.ifspsaocarlos.agenda.adapter.ContactArrayAdapter
import br.edu.ifspsaocarlos.agenda.data.ContactDAO
import br.edu.ifspsaocarlos.agenda.model.Contact

open class BaseActivity : AppCompatActivity() {

    private val list by lazy { findViewById<ListView>(R.id.listView) }
    private var adapter: ContactArrayAdapter? = null

    private val contactDAO: ContactDAO by lazy { ContactDAO(this@BaseActivity.applicationContext) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        list.onItemClickListener = OnItemClickListener { adapterView, _, arg2, _ ->
            val contact = adapterView.adapter.getItem(arg2) as Contact
            val intent = Intent(applicationContext, DetailActivity::class.java)
            intent.putExtra("contact", contact)
            startActivityForResult(intent, 0)
        }

        registerForContextMenu(list)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)

        val searchManager = getSystemService(SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.searchContact).actionView as SearchView
        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.setIconifiedByDefault(true)

        return true
    }

    override fun onCreateContextMenu(
        menu: ContextMenu, v: View,
        menuInfo: ContextMenuInfo
    ) {
        super.onCreateContextMenu(menu, v, menuInfo)
        val menuInflater = menuInflater
        menuInflater.inflate(R.menu.menu_context, menu)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as AdapterContextMenuInfo? ?: return false
        val adapter = list.adapter as ContactArrayAdapter
        val contact = adapter.getItem(info.position)

        if (item.itemId == R.id.delete_item) {
            contactDAO.deleteContact(contact!!)
            Toast.makeText(applicationContext, "Removido com sucesso", Toast.LENGTH_SHORT).show()
            buildListView()
            return true
        }
        return super.onContextItemSelected(item)
    }

    protected fun buildListView() {
        val values = contactDAO.searchAllContacts()
        adapter = ContactArrayAdapter(this, values)
        list.adapter = adapter
    }

    protected fun buildSearchListView(query: String) {
        val values = contactDAO.searchContact(query)
        adapter = ContactArrayAdapter(this, values)
        list.adapter = adapter
    }
}
